### PR TITLE
Authorization rules

### PR DESCRIPTION
**Pull request by the-faizmohammad**
## Pull Request Summary: Enhanced User Roles and Authorization

This pull request implements several features related to user roles and authorization:

**1. User Roles and CanCanCan Integration:**

* Installs CanCanCan gem for authorization logic.
* Adds a `role` column to the `users` table with a migration.
* Defines CanCanCan abilities for users based on their role (`user` or `admin`).

**2. Post Deletion with Authorization:**

* Implements functionality to delete posts.
* Adds a "Delete" button to the post view.
* Uses CanCanCan to restrict post deletion to the post owner or users with the `admin` role.

**3. Comment Deletion with Authorization:**

* Implements functionality to delete comments.
* Adds a "Delete" button to the comment view.
* Uses CanCanCan to restrict comment deletion to the comment owner or users with the `admin` role.

**This pull request enhances user roles and empowers users with appropriate permissions while ensuring security through CanCanCan integration.**

**Please review the changes and provide feedback.**

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    can :manage, Post, author_id: user.id
+    can :create, Comment
+    can :manage, Comment, user_id: user.id
+    can :destroy, Like, user_id: user.id
+
+    # Abilities for admin users
+    can :manage, :all if user.admin?
+  end
+end
```
Comment (danmainah) [inline app/models/ability.rb:19]: - [ ] @Team Great work adding the Ability class to add authorization, however, I could not find the CanCan gem in your gemfile which I suspect is leading to the error below, CanCan gem is part of the Ability class and is used for authorization in Ruby on Rails applications. Kindly consider adding the Gem.
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @the-faizmohammad and @ReemMohamedAbdelfatah , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️ Added the ability class and define the roles of each user
✔️  Added method admin in the user model to define who is the admin.


# Required Changes ♻️

- [ ] @Team Kindly consider ensuring the user is able to delete their own posts and comments, by adding the delete functionality to test if the ability class is working correctly.

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/views/posts/show.html.erb [hunk 2]
```diff
@@ -14,7 +14,15 @@
             <p class="each-commnet">No comments yet!</p>
         <% else %>
             <% @post.comments.each do |comment| %>
-            <p class="each-comment">@<%= comment.user.name %>: <%= comment.text %></p>
+                <div class='comment-delete'>
+                 <p class="each-comment">@<%= comment.user.name %>: <%= comment.text %></p>
+
+                 <% if can? :destroy, comment %>
+                   <%= form_with(model: comment, url: user_post_comment_path(@post.author, @post, id: comment.id), method: :delete, id: 'delete-comment-form') do |form| %>
```
Comment (skyv26) [inline app/views/posts/show.html.erb:21]: - [ ] Well done @the-faizmohammad and @ReemMohamedAbdelfatah, you did a great job by implementing the changes as suggested by your previous code reviewer, but I noticed that, when I click on a specific post to see the comments, likes, `delete` comment etc, I got an error for that specific post and the reason behind the error is the `wrong` path that you gave for the `URL`. The path should be `user_post_comments_path` not ❌ `user_post_comment_path`. 

image


- [ ] After solving the above issue, I got another issue. I saw that when I click on the delete button to remove a particular comment, it deletes the comment from that place and `renders` the comment again and again on the page, I suspect that your `delete` comment action is not working, which means you button click event only refresh the page but not deleting the comment from the database. Kindly fix the behaviour and make sure that your application should work properly and in the expected way. Remember that only the admin should be allowed to `delete` any person's comments or `posts`. Rest the `normal` login user should only be able to delete its own `comments` or `posts`.

[Screencast from 10-02-24 01:11:21 AM IST.webm](https://github.com/the-faizmohammad/BlogApp/assets/66476329/b831d50d-1fd8-4add-bcf5-d09f9fe79261)

**Review by skyv26 — CHANGES_REQUESTED** [review]
## Hi @the-faizmohammad and @ReemMohamedAbdelfatah, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## STATUS: REQUIRE CHANGES 🔴 




## Highlights
- Added the `cancancan` library ✅ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

#### Cheers, and Happy coding!👏👏👏

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by skyv26 — APPROVED** [review]
## Hi @the-faizmohammad and @ReemMohamedAbdelfatah, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## STATUS: APPROVED 🟢 



## Highlights
- Impressive work and implemented the missing feature as suggested previously ✅ 

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**